[MSE][GStreamer] Initial support for multiple video tracks on MSE

eocanha opened the pull request:
#### dffed01e1b6e5290e880e9cfb2c27cdabfa83a42
<pre>
[MSE][GStreamer] Initial support for multiple video tracks on MSE
<a href="https://bugs.webkit.org/show_bug.cgi?id=298245">https://bugs.webkit.org/show_bug.cgi?id=298245</a>

Reviewed by Philippe Normand.

Multiple video tracks in MSE are (optionally) supported by the spec[1],
but have never been used in practice in WebKitGTK/WPE. Creating more
than one video SourceBuffer causes a not-linked pipeline error in
WebKitMediaSrc. It would be nice to have a basic support of that use
case, since it&apos;s apparently required for the YouTube EME tests.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1549">https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1549</a>

This patch sanitizes the video track selection state on creation (only
one video track can be active at once, all of them appeared as active on
creation before), mirrors the selection state from the video tracks
known to the HTMLMediaElement to the ones knows by the player private
(they&apos;re a different set in the case of MSE), and adds support for
unregistered streams in WebKitMediaSrc (the ones causing the not-linked
error in the first place), just avoiding pushing buffers to pads with
unlinked steams.

This patch only provides an initial support. It doesn&apos;t address complex
scenarios such as the init segment and data reenqueues that should be
needed if the switch happens after the init segment and data have been
appended (and enqueued) and the track selection change happens after
that. In such cases, the pipeline is refactored but, since the needed
data hasn&apos;t reached the video decoder on time (because the decoder just
didn&apos;t exist at enqueue time), only future samples will be enqueued from
that time on and decoding artifacts will happen.

A layout test has been added to test multiple video track basic support,
ensuring that the track ids are properly generated and that only
the first track is selected.

[1] <a href="https://w3c.github.io/media-source/#dfn-sourcebuffer-configuration">https://w3c.github.io/media-source/#dfn-sourcebuffer-configuration</a>

Layout test based on code provided by: Filipe Norte &lt;filipe_norte@comcast.com&gt;

* LayoutTests/media/media-source/content/video320x180-101-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-init.mp4: Added.
* LayoutTests/media/media-source/content/video320x180-102-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-init.mp4: Added.
* LayoutTests/media/media-source/content/video320x180-103-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-init.mp4: Added.
* LayoutTests/media/media-source/media-source-multiple-video-tracks-expected.txt: Added.
* LayoutTests/media/media-source/media-source-multiple-video-tracks.html: Test that created 3 different SourceBuffers, loads videos with 3 different track ids, and checks that the proper ids are generated and that only the first video trac&gt;
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::addVideoTrackToElement): Select the first video track added and unselect the next ones to ensure that only one video track is active at a given time after adding several ones.
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::setSelected): Propagate the &quot;selected&quot; state to the mirror track that the player private may have.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::mirrorEnabledVideoTrackIfNeeded): Propagate the &quot;selected&quot; state from a given VideoTrackPrivateGStreamer (typically one know by HTMLMediaElement) to the internal VideoTrackPrivateGStreamer managed by &gt;
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added mirrorEnabledVideoTrackIfNeeded() empty implementation.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h: Added mirrorEnabledVideoTrackIfNeeded().
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(WebKitMediaSrcPrivate::streamById): Support asking for streams not registered in this element. Log an info instead of asserting when not present.
(webKitMediaSrcTearDownStream): Gracefully bail out if the Stream is not registered in this element.
(webKitMediaSrcLoop): Ditto. Just skip the buffer push without triggering an error.
(webKitMediaSrcFlush): Ditto.

Canonical link: <a href="https://commits.webkit.org/299700@main">https://commits.webkit.org/299700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8450477682fa208c83f5699e06539f0455621b76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125922 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47908 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90870 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60172 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31948 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107274 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71385 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25379 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69566 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101415 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128884 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46558 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35270 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99467 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25261 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44732 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22758 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43122 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46420 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52126 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45886 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62906009a5020d3e7a4350f2b48f8b864847c039. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50187

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -206,6 +206,12 @@ class MediaPlayerPrivateGStreamer
+    // On MSE, the player holds its own set of tracks, independent from the ones SourceBuffer
+    // reported to HTMLMediaElement. We need to synchronize the enabled status of the player
+    // mirror when the element one changed. Fortunately, both share the same trackId.
+    void mirrorEnabledVideoTrackIfNeeded(const VideoTrackPrivateGStreamer& originalVideoTrackPrivate);
```
As this is MSE-specific, maybe make it virtual and implement in the MSE player?

eocanha committed:
[MSE][GStreamer] Initial support for multiple video tracks on MSE
https://bugs.webkit.org/show_bug.cgi?id=298245

Reviewed by Philippe Normand.

Multiple video tracks in MSE are (optionally) supported by the spec[1],
but have never been used in practice in WebKitGTK/WPE. Creating more
than one video SourceBuffer causes a not-linked pipeline error in
WebKitMediaSrc. It would be nice to have a basic support of that use
case, since it's apparently required for the YouTube EME tests.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1549

This patch sanitizes the video track selection state on creation (only
one video track can be active at once, all of them appeared as active on
creation before), mirrors the selection state from the video tracks
known to the HTMLMediaElement to the ones knows by the player private
(they're a different set in the case of MSE), and adds support for
unregistered streams in WebKitMediaSrc (the ones causing the not-linked
error in the first place), just avoiding pushing buffers to pads with
unlinked steams.

This patch only provides an initial support. It doesn't address complex
scenarios such as the init segment and data reenqueues that should be
needed if the switch happens after the init segment and data have been
appended (and enqueued) and the track selection change happens after
that. In such cases, the pipeline is refactored but, since the needed
data hasn't reached the video decoder on time (because the decoder just
didn't exist at enqueue time), only future samples will be enqueued from
that time on and decoding artifacts will happen.

A layout test has been added to test multiple video track basic support,
ensuring that the track ids are properly generated and that only
the first track is selected.

[1] https://w3c.github.io/media-source/#dfn-sourcebuffer-configuration

Layout test based on code provided by: Filipe Norte <filipe_norte@comcast.com>

* LayoutTests/media/media-source/content/video320x180-101-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-101-init.mp4: Added.
* LayoutTests/media/media-source/content/video320x180-102-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-102-init.mp4: Added.
* LayoutTests/media/media-source/content/video320x180-103-1.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-2.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-3.m4s: Added.
* LayoutTests/media/media-source/content/video320x180-103-init.mp4: Added.
* LayoutTests/media/media-source/media-source-multiple-video-tracks-expected.txt: Added.
* LayoutTests/media/media-source/media-source-multiple-video-tracks.html: Test that created 3 different SourceBuffers, loads videos with 3 different track ids, and checks that the proper ids are generated and that only the first video trac>
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::addVideoTrackToElement): Select the first video track added and unselect the next ones to ensure that only one video track is active at a given time after adding several ones.
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::setSelected): Propagate the "selected" state to the mirror track that the player private may have.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::mirrorEnabledVideoTrackIfNeeded): Propagate the "selected" state from a given VideoTrackPrivateGStreamer (typically one know by HTMLMediaElement) to the internal VideoTrackPrivateGStreamer managed by >
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added mirrorEnabledVideoTrackIfNeeded() empty implementation.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h: Added mirrorEnabledVideoTrackIfNeeded().
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(WebKitMediaSrcPrivate::streamById): Support asking for streams not registered in this element. Log an info instead of asserting when not present.
(webKitMediaSrcTearDownStream): Gracefully bail out if the Stream is not registered in this element.
(webKitMediaSrcLoop): Ditto. Just skip the buffer push without triggering an error.
(webKitMediaSrcFlush): Ditto.

Canonical link: https://commits.webkit.org/299700@main
